### DISCUSSION TITLE
Dark Mode - some success to stopping the flash

### DISCUSSION THREAD

**Discussion by Brian-McBride** [mantinedev/mantine#2319] [Feedback]
## Overview

Mantine has a good start on the Dark mode support for Next.js. However, if you try and set up something that defaults to the user's browser preferences, then you run the risk of a flash of light.

We can think of it:
```mermaid
  graph LR;
Light --> Media --> Cookie
```

Where we want them to cascade and overwrite the `ColorScheme` value (`light` or `dark`)

## Problem

Next.js will pre-render the HTML on the server and send it to the client for the first paint. While that is happening, the actual JS code (like the react runtime and your needed libs for that page) is loaded and hydrated. 

If I log out the server's `colorScheme` value I'll see something like this:
```js
{ colorScheme: 'light' }
```

At the same time, let's say I either have dark mode on my browser or I've previously set a cookie for a dark theme via a toggle switch. The client will have:
```js
{ colorScheme: 'dark' }
```

Thus, there will be a split moment when the background flashes white. 

## Partial Fix

I ended up going for a simpler solution using the npm lib `next-themes`.
It is just another context provider with a bit of helper logic. The source is pretty clear. 

**`_app.tsx`**
```ts
import { AppProps } from 'next/app';
import Head from 'next/head';
import { ProviderWraps } from '../components/ProviderWraps';
import '../styles/global.css';

function NextApp({ Component, pageProps }: AppProps) {
  return (
    <>
      <Head>
        <meta
          name="viewport"
          content="width=device-width, minimum-scale=1, maximum-scale=1, initial-scale=1, viewport-fit=cover"
        />
        <title>Sydekick Tasks</title>
      </Head>
      <ProviderWraps>
        <Component {...pageProps} />
      </ProviderWraps>
    </>
  );
}

export default NextApp;
```

I'm using a component to handle all my providers. In this example, I'll show a provider with `react-query`.

**`components/ProviderWraps.ts`**
```ts
import { ColorScheme, ColorSchemeProvider, MantineProvider } from '@mantine/core';
import { useHotkeys } from '@mantine/hooks';
import { NotificationsProvider } from '@mantine/notifications';
import { getCookie, setCookie } from 'cookies-next';
import { ThemeProvider, useTheme as useNextTheme } from 'next-themes';
import { PropsWithChildren, useEffect, useRef, useState } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MAX_NOTIFICATIONS } from '...your.path../settings';
import { mantineTheme } from '..your.path../theme';

function generateClients() {
  const queryClient = new QueryClient();
  return { queryClient };
}

function getScheme() {
  const userScheme = getCookie('color-scheme') as ColorScheme;
  return userScheme === 'dark' ||
    (userScheme !== 'light' &&
      window.matchMedia('(prefers-color-scheme: dark)').matches)
    ? 'dark'
    : 'light';
}

export function ProviderWraps({ children }: PropsWithChildren) {
  //* React Query *//
  const { queryClient } = useRef(generateClients()).current;

  const [colorScheme, setColorScheme] = useState<ColorScheme | null>();

  useEffect(() => {
    if (!colorScheme) {
      const mode = getScheme();
      console.log(`Setting color scheme to ${mode}`);
      setColorScheme(mode);
      document.documentElement.style.setProperty(
        '--on-first-render',
        'visible'
      );
    }
    if (colorScheme === 'dark') {
      document.documentElement.classList.add('dark');
      document.getElementsByTagName('BODY')[0].classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
      document.getElementsByTagName('BODY')[0].classList.remove('dark');
    }
  }, [colorScheme]);

  const toggleColorScheme = (value?: ColorScheme) => {
    const mode = value || (colorScheme === 'dark' ? 'light' : 'dark');
    setCookie('color-scheme', mode);
    setColorScheme(mode);
  };
  useHotkeys([['mod+J', () => toggleColorScheme()]]);

  return (
      <QueryClientProvider client={queryClient}>
        <ColorSchemeProvider
          toggleColorScheme={toggleColorScheme}
          colorScheme={colorScheme || 'light'}
        >
          <MantineProvider
            withGlobalStyles
            withNormalizeCSS
            theme={{ ...mantineTheme, colorScheme: colorScheme || 'light' }}
          >
            <NotificationsProvider limit={MAX_NOTIFICATIONS}>
              <main className={'app-wrap'}>{children}</main>
            </NotificationsProvider>
          </MantineProvider>
        </ColorSchemeProvider>
      </QueryClientProvider>
  );
}
```

You'll need to add a script to your `_document` file:
```js
function getCookie(name) {
  var value = `; ${document.cookie}`;
  var parts = value.split(`; ${name}=`);
  if (parts.length === 2) return parts.pop().split(';').shift();
}
var colorScheme = getCookie('color-scheme');
var prefersDark = window.matchMedia('(prefers-color-scheme: dark)').matches;
if (colorScheme === 'dark' || (colorScheme !== 'light' && prefersDark)) {
  document.documentElement.classList.add('dark')
  document.getElementsByTagName("BODY")[0].classList.add('dark')
  document.documentElement.style.setProperty('--on-first-render', 'hidden');
} else {
  document.documentElement.classList.remove('dark')
  document.getElementsByTagName("BODY")[0].classList.remove('dark')
  document.documentElement.style.setProperty('--on-first-render', 'visible');
}
```

You use it sort of like `<Script src="scripts/dark-mode.js" strategy="beforeInteractive" />` with `import Script from 'next/script';` 

And some CSS vars to handle it all
```css
:root {
  --on-first-render: visible;
}
.app-wrap {
  visibility: var(--on-first-render, visible);
}
```

As @rtivital  mentioned below, I'm hiding the content until the theme is loaded. But only in the case of the dark theme. If it is light, I let that just start visible. This is the best compromise I can make and still avoid flash.

However, on SSR you could load in the `color-scheme` cookie in `getInitialProps` like in the Mantine docs, adjust the `_document` script a bit to handle that use case, or just remove the visibility variables. 

## Rationalization

Flash or immediate paint. That is the question here. Remember, the process is:

1. Server renders a static HTML page
2. Client downloads static page and displays it __(the light theme)__
3. Client is downloading and finishing the Javascript bundle
4. Client hydrates that code, and now regular React is in place. __(now replaces with the actual theme)__
5. Navigation from this point on uses the hydrated state (unless you don't use the Next router)

If you currently have a super fast flash. That's the time it takes to display the first paint page and then the hydrated page. That's pretty darn quick in most cases. So, dark mode users have to wait that extra time for visibility. Light mode visitors don't have to wait, next.js is normal. 

Because the element is hidden, you will get the SSR/SSG benefits of SEO. Although I'm not sure how much that will affect Google's Bots when indexing a hidden page. It could lower the score.

**Comment by rtivital** [comment]
Hi @Brian-McBride, you are just hiding the content before js bundle loads, for me, it defeats the purpose of ssr, you can switch to vite and have the same results but without any tricks.

**Reply by Brian-McBride** [reply]
@rtivital  I've updated my solution. It now won't hide the `light` theme, but it will hide the `dark` theme until after hydration. I don't believe it completely defeats the point of SSR (although I'm using SSG). 

I wonder how many people have a flash of white that lasts more than a fraction of a second? In the end, settling on the dark mode is when the site reaches 'first interactive'. 
I don't understand how Vite solves any of this though. Pre-generated pages have no way to know a user's theme preference. 

The real solution here is to have CSS variables driving the root theme everywhere. Then they could just be replaced as needed way up in the `_document` above all the react code.

**Comment by Brian-McBride** [comment]
It looks like MUI is trying something similar with CSS variables:  https://github.com/mui/material-ui/issues/27651

**Comment by Bladeyy** [comment]
Nice. I'm also currently making do with next-themes package to deal with FART and achieve perfect darkmode in nextjs. Implementation is pretty straight forward if you read their docs. All you need is to wrap next-themes ThemeProvider around MantineProvider and use their resolvedTheme object to set MantineColorScheme. Would be cool if Mantine has in-house implementation to acheive that though. But anyway that's not a must need feature. May be we can see it in the future versions.
